cryosparc/api.py

```diff
@@ -0,0 +1,409 @@
+import json
+import re
+import urllib.parse
+import warnings
+from contextlib import contextmanager
+from enum import Enum
+from typing import Any, Dict, Iterator, Optional, Tuple, TypedDict, Union
+
+import httpx
+
+from . import registry
+from .errors import APIError
+from .json_util import api_default, api_object_hook
+from .models.auth import Token
+from .stream import Streamable
+
+_BASE_RESPONSE_TYPES = {"string", "integer", "number", "boolean"}
+
+Auth = Union[str, tuple[str, str]]
+"""
+Auth token or email/password.
+"""
+
+
+class APIRequest(TypedDict):
+    params: Dict[str, Any]
+    headers: Dict[str, str]
+    content: Any
+    data: Optional[Dict[str, Any]]  # form data
+    files: Optional[Dict[str, Any]]
+
+
+class APINamespace:
+    """
+    Collection of API methods that call a certain namespace, e.g., only the
+    methods under http://master:39004/pipelines/
+    """
+
+    _client: httpx.Client
+
+    def __init__(self, http_client: httpx.Client):
+        self._client = http_client
+
+    def _set_headers(self, update: Dict[str, str]):
+        """For testing only, reset client headers"""
+        self._client.headers.update(update)
+
+    def _prepare_request_stream(self, streamable):
+        return streamable.stream()
+
+    def _construct_request(self, _path: str, _schema, *args, **kwargs) -> Tuple[str, APIRequest]:
+        args = list(args)
+        query_params = {}
+        func_name = _schema["summary"]
+        headers = {}
+        client_headers = {h for h in self._client.headers.keys()}
+        request_body = None
+        data = None
+        files = None
+
+        for param_schema in _sort_params_schema(_path, _schema.get("parameters", [])):
+            # Compile function params
```

Comment:
Interpreting params is a bit different. Path params must now be specified positionally. Query and header params must now be specified as keyword args.